firstpartydata/first_party_data.go

```diff
@@ -0,0 +1,405 @@
+				newUser.Ext = extData
+			}
+		}
+		if len(openRtbGlobalFPD[userDataKey]) > 0 {
+			newUser.Data = openRtbGlobalFPD[userDataKey]
+		}
+	}
+	if fpdConfigUser != nil {
+		//apply bidder specific fpd if present
+		newUser, err = mergeUsers(&newUser, fpdConfigUser)
+	}
+
+	return &newUser, err
+}
+
+func mergeUsers(original *openrtb2.User, fpdConfigUser *openrtb2.User) (openrtb2.User, error) {
+
+	var err error
+	newUser := *original
+	newUser.Keywords = fpdConfigUser.Keywords
+	newUser.Gender = fpdConfigUser.Gender
+	newUser.Yob = fpdConfigUser.Yob
+
+	if len(fpdConfigUser.Ext) > 0 {
+		if len(original.Ext) > 0 {
+			newUser.Ext, err = jsonpatch.MergePatch(original.Ext, fpdConfigUser.Ext)
+		} else {
+			newUser.Ext = fpdConfigUser.Ext
+		}
+	}
+
+	return newUser, err
+}
+
+func resolveSite(fpdConfigSite *openrtb2.Site, bidRequestSite *openrtb2.Site, globalFPD map[string][]byte, openRtbGlobalFPD map[string][]openrtb2.Data, hasGlobalFPD bool, bidderName string) (*openrtb2.Site, error) {
+
+	if bidRequestSite == nil && fpdConfigSite == nil {
+		return nil, nil
+	}
+	if bidRequestSite == nil && fpdConfigSite != nil {
+		return nil, fmt.Errorf("incorrect First Party Data for bidder %s: Site object is not defined in request, but defined in FPD config", bidderName)
+	}
+
+	newSite := *bidRequestSite
+	var err error
+
+	if hasGlobalFPD {
+		//apply global fpd
+		if len(globalFPD[siteKey]) > 0 {
+			extData := buildExtData(globalFPD[siteKey])
+			if len(newSite.Ext) > 0 {
+				newSite.Ext, err = jsonpatch.MergePatch(newSite.Ext, extData)
+			} else {
+				newSite.Ext = extData
+			}
+		}
+		if len(openRtbGlobalFPD[siteContentDataKey]) > 0 {
+			if newSite.Content != nil {
+				contentCopy := *newSite.Content
+				contentCopy.Data = openRtbGlobalFPD[siteContentDataKey]
+				newSite.Content = &contentCopy
+			} else {
+				newSite.Content = &openrtb2.Content{Data: openRtbGlobalFPD[siteContentDataKey]}
+			}
+		}
+	}
+
+	if fpdConfigSite != nil {
+		//apply bidder specific fpd if present
+		//result site should have ID or Page, fpd becomes incorrect if it overwrites page to empty one and ID is empty in original site
+		if fpdConfigSite.Page == "" && newSite.Page != "" && newSite.ID == "" {
+			return nil, fmt.Errorf("incorrect First Party Data for bidder %s: Site object cannot set empty page if req.site.id is empty", bidderName)
+
+		}
+		newSite, err = mergeSites(&newSite, fpdConfigSite)
+	}
+	return &newSite, err
+
+}
+
+func mergeSites(originalSite *openrtb2.Site, fpdConfigSite *openrtb2.Site) (openrtb2.Site, error) {
+
+	var err error
+	newSite := *originalSite
+
+	newSite.Name = fpdConfigSite.Name
+	newSite.Domain = fpdConfigSite.Domain
+	newSite.Cat = fpdConfigSite.Cat
+	newSite.SectionCat = fpdConfigSite.SectionCat
+	newSite.PageCat = fpdConfigSite.PageCat
+	newSite.Page = fpdConfigSite.Page
+	newSite.Search = fpdConfigSite.Search
+	newSite.Keywords = fpdConfigSite.Keywords
+	newSite.Ref = fpdConfigSite.Ref
+
+	if len(fpdConfigSite.Ext) > 0 {
```

Comment:
As discussed, `mergeSites`, `mergeApps` and `mergeUsers` should add all ATTRs that don't map to a root.ATTR in `{app/site/user}.ext.data.ATTR`.
For example, the ORTB2 site object contains fields `id`, `privacypolicy` and `mobile` which do not explicitly map to `site.id`, `site.privacypolicy` and `site.mobile` according to the FPD requirements. If these fields exist in the bidderconfig, they should be placed at `site.ext.data.id`, `site.ext.data.privacypolicy` and `site.ext.data.mobile` respectively.

Reply:
Pushed initial implementation. Unit tests - to fix. Also need to fix extra escape characters when converting json.RawMessage to string.
@bsardo if you will tests it, please replace `\\\"` to empty string first in request from debug logs :)

Reply:
Fixed. Should work according to the way you described